File: java/simple-linked-list/src/main/java/SimpleLinkedList.java
import java.lang.reflect.Array;
import java.util.NoSuchElementException;

class SimpleLinkedList<T> {
    private Node<T> head;
    private int size;

    // Node class for the linked list
    private static class Node<T> {
        private T data;
        private Node<T> next;

        Node(T data) {
            this.data = data;
            this.next = null;
        }
    }

    // Default constructor
    SimpleLinkedList() {
        this.head = null;
        this.size = 0;
    }

    // Constructor from array
    SimpleLinkedList(T[] values) {
        this();
        if (values != null) {
            // Add elements in reverse order to maintain LIFO behavior
            for (int i = values.length - 1; i >= 0; i--) {
                push(values[i]);
            }
        }
    }

    // Add element to the front of the list
    void push(T value) {
        Node<T> newNode = new Node<>(value);
        newNode.next = head;
        head = newNode;
        size++;
    }

    // Remove and return the element from the front of the list
    T pop() {
        if (head == null) {
            throw new NoSuchElementException("Cannot pop from an empty list");
        }
        
        T value = head.data;
        head = head.next;
        size--;
        return value;
    }

    // Reverse the list
    void reverse() {
        if (head == null || head.next == null) {
            return; // Empty list or single element list doesn't need reversal
        }
        
        Node<T> prev = null;
        Node<T> current = head;
        Node<T> next;
        
        while (current != null) {
            next = current.next;  // Store next node
            current.next = prev;  // Reverse the link
            prev = current;       // Move prev to current
            current = next;       // Move current to next
        }
        
        head = prev; // Update head to the new front (previously the last node)
    }

    // Convert list to array
    @SuppressWarnings("unchecked")
    T[] asArray(Class<T> clazz) {
        T[] array = (T[]) Array.newInstance(clazz, size);
        
        Node<T> current = head;
        for (int i = 0; i < size; i++) {
            array[i] = current.data;
            current = current.next;
        }
        
        return array;
    }

    // Return the size of the list
    int size() {
        return size;
    }
}

File: javascript/simple-linked-list/simple-linked-list.js
//
// This is only a SKELETON file for the 'Simple Linked List' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Element {
  constructor(value) {
    this._value = value;
    this._next = null;
  }

  get value() {
    return this._value;
  }

  get next() {
    return this._next;
  }

  set next(element) {
    this._next = element;
  }
}

export class List {
  constructor(values = []) {
    this._length = 0;
    this._head = null;
    
    // Initialize from array if provided
    if (values.length > 0) {
      for (let i = 0; i < values.length; i++) {
        this.add(new Element(values[i]));
      }
    }
  }

  add(element) {
    if (this._head) {
      element.next = this._head;
    }
    this._head = element;
    this._length++;
  }

  get length() {
    return this._length;
  }

  get head() {
    return this._head;
  }

  toArray() {
    const result = [];
    let current = this._head;
    
    while (current) {
      result.push(current.value);
      current = current.next;
    }
    
    return result;
  }

  reverse() {
    const values = this.toArray();
    const newList = new List();
    
    // Create a new list with elements in the reverse order
    // Since our add method adds to the front, we need to iterate in the original order
    for (let i = 0; i < values.length; i++) {
      newList.add(new Element(values[i]));
    }
    
    return newList;
  }
}
